### DISCUSSION TITLE
Implement window.location.href only using wasm_bindgen without using web_sys::Window.

### DISCUSSION THREAD

**Discussion by dataangel** [wasm-bindgen/wasm-bindgen#3771] [Q&A]
Seek help.
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = ["window", "location"])]
    #[wasm_bindgen(js_name = "href")]
    fn window_location_href() -> JsValue;
}

#[wasm_bindgen]
pub fn get_window_location_href() -> String {
    let href_value = window_location_href();

    // 尝试将 href 转换为字符串
    let href = href_value
        .as_string()
        .unwrap_or_else(|| String::from("Error getting window.location.href"));

    href
}
window.location.href is no function

**Comment by dataangel** [comment]
export function __wbg_href_0286eba8d0d35cc2() { return logError(function () {
    const ret = window.location.href();<-----
    return addHeapObject(ret);
}, arguments) };
this

**Comment by daxpedda** [comment]
I don't think `wasm-bindgen` supports global properties. Might be considered a bug, because the `getter` attribute is just ignored.
You will have to go through `js_sys::global()` unless I'm missing something.

**Comment by dataangel** [comment]
getter not working

**Comment by ghost** [comment]
```rust 
#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen]
    pub type WindowLocation;

    #[wasm_bindgen]
    pub type Window;

    #[wasm_bindgen(js_name = "window")]
    pub static WINDOW: Window;

    #[wasm_bindgen(method, getter, js_name = "location")]
    pub fn get_location(this: &Window) -> WindowLocation;

    #[wasm_bindgen(method, getter, js_name = "href")]
    pub fn get_href(this: &WindowLocation) -> JsValue;
}

#[wasm_bindgen]
pub fn retrieve_href() -> JsValue {
    WINDOW.get_location().get_href()
}
```

By the way, [`web_sys::Window`](https://rustwasm.github.io/wasm-bindgen/api/src/web_sys/features/gen_Window.rs.html#14) uses wasm_bindgen. It looks like it does it differently though.

```rust
#[wasm_bindgen]
extern "C" {
    # [wasm_bindgen (extends = EventTarget , extends = :: js_sys :: Object , js_name = Window , typescript_type = "Window")]
    #[derive(Debug, Clone, PartialEq, Eq)]
    pub type Window;
}
```

**Reply by daxpedda** [reply]
I wasn't aware `#[wasm_bindgen]` supported `static`, nice!

```rust
#[wasm_bindgen]
extern "C" {
    type Location;

    #[wasm_bindgen(js_name = "location")]
    static LOCATION: Location;

    #[wasm_bindgen(method, getter, js_name = "href")]
    fn href(this: &Location) -> String;
}

web_sys::console::log_1(&LOCATION.href().into());
```
Is probably the shortest way to do this, as `Window` is already accessible through `web_sys::window()`.

**Reply by ghost** [reply]
There's lots of ways to write bindings. 

```rust
#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_name = "href", js_namespace=["window", "location"])]
    pub static HREF: JsValue;
}

#[wasm_bindgen]
pub fn retrieve_href() -> JsValue {
    JsValue::from(HREF.as_string())
}
```

I spent more time than I probably should have on `web_sys` and `js_sys`. I wish I would have read [wasm_bindgen book](https://rustwasm.github.io/wasm-bindgen/reference/attributes/on-js-imports/index.html) carefully at first because I think it's easier than trying to use stuff like `Reflect`.
